### DISCUSSION TITLE
Exception raised when hitting Ctrl-C in uvicorn >= 0.29.0 if started programmatically

### DISCUSSION THREAD

**Discussion by GianlucaFicarelli** [Kludex/uvicorn#2368] [Potential Issue]
There is a similar issue https://github.com/encode/uvicorn/issues/2289, but it's marked as solved so I'm not sure that the problem is the same.
I'm opening the discussion with my example that can be used to reproduce the error.
In a virtualenv on mac with:
- python 3.12
- uvicorn==0.30.1
- fastapi==0.111.0

I have these files:

app.py:
```python
from fastapi import FastAPI
app = FastAPI(title="test")
```

main.py:
```python
import uvicorn
if __name__ == "__main__":
    uvicorn.run("app:app")
```

If I start uvicorn with `uvicorn app:app`, and hit ctrl-c, I get what I expect:
```
INFO:     Started server process [88323]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [88323]

Aborted!
```
But if I start uvicorn programmatically with `python -m main` and hit ctrl-c, then an exception is propagated:
```
INFO:     Started server process [88391]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
^CINFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [88391]
Traceback (most recent call last):
  File "/path/to/.pyenv/versions/3.12.1/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/private/tmp/src/main.py", line 5, in <module>
    uvicorn.run( "app:app")
  File "/private/tmp/venv/lib/python3.12/site-packages/uvicorn/main.py", line 577, in run
    server.run()
  File "/private/tmp/venv/lib/python3.12/site-packages/uvicorn/server.py", line 65, in run
    return asyncio.run(self.serve(sockets=sockets))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/path/to/.pyenv/versions/3.12.1/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/path/to/.pyenv/versions/3.12.1/lib/python3.12/asyncio/runners.py", line 123, in run
    raise KeyboardInterrupt()
KeyboardInterrupt

❯ echo $?
130
```

I verified that the issue wasn't present in uvicorn 0.28.1, and it appeared in 0.29.0, so it's probably caused by https://github.com/encode/uvicorn/pull/1600

**Comment by maxfischer2781** [comment]
Can you clarify why you consider this an issue? [Raising `KeyboardInterrupt` on CTRL+C is the normal Python behaviour](https://docs.python.org/3/library/exceptions.html#KeyboardInterrupt), and things break if this is suppressed completely (see #1579). #1600 *deliberately* added this back in.

If you want to suppress the exception, you can do so the usual way:
```
import uvicorn
if __name__ == "__main__":
    try:
        uvicorn.run("app:app")
    except KeyboardInterrupt:
        pass
```

**Reply by Kludex** [reply]
That's not practical... 😔

**Reply by maxfischer2781** [reply]
Do you think a flag or separate method would be more practical to select the behaviour? At least maintenance-wise, it would be rather simple to offer two versions of `uvicorn.run`.

An alternative would be for `uvicorn.run` to trim the `KeyboardInterrupt` traceback. The details of `asyncio` and `uvloop`, and probably `uvicorn` itself, are pointless to know and always the same. Would a shorter traceback like this be a compromise?
```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/private/tmp/src/main.py", line 5, in <module>
    uvicorn.run( "app:app")
  File "/private/tmp/venv/lib/python3.12/site-packages/uvicorn/main.py", line 1337, in run
    raise KeyboardInterrupt() from None
KeyboardInterrupt
```
I'm not sure if it's possible to remove the entire traceback from inside `uvicorn`.

**Reply by GianlucaFicarelli** [reply]
To clarify, I'm considering this an issue because I would expect the same behaviour when uvicorn is executed [programmatically](https://www.uvicorn.org/#running-programmatically) (with just more flexibility, for example to configure logging).
The message printed is in both the cases says "Press CTRL+C to quit" so I won't expect an exception to be raised by default.
But having a parameter as you suggested seems reasonable to me.

**Comment by maxfischer2781** [comment]
See #2383 as a draft for quieter/silent quitting. Feedback is welcome. I'm specifically not sure whether the shorter or silent shutdown should be the default.

**Reply by maxfischer2781** [reply]
Closed in favour of #2384.

**Comment by GianlucaFicarelli** [comment]
Commenting here, although it's related also to the PRs:

https://github.com/encode/uvicorn/pull/1600
https://github.com/encode/uvicorn/pull/2383

When uvicorn is used programmatically, and it's part of a bigger async application, instead of re-raising signals, wouldn't be cleaner and simpler to let the user decide to:

- start uvicorn without signal handler
- register its own signal handlers, that should be responsible to:
  - stop uvicorn
  - stop any other task if needed
  - exit the program

This can be already done by calling the private method `_serve()` instead of `serve()`, but it would be even better to expose a public API.

In this way, it's still possible to integrate uvicorn into bigger applications without changing the behaviour of `uvicorn.run()`.

Simplified working example:

```python
import asyncio
import signal

import uvicorn
from fastapi import FastAPI

HANDLED_SIGNALS = (
    signal.SIGINT,
    signal.SIGTERM,
)

app = FastAPI(title="test")


class Sleeper:
    def __init__(self, sleep: float = 10) -> None:
        self.should_exit = False
        self.sleep = sleep

    async def run_forever(self) -> None:
        while not self.should_exit:
            print(f"Sleeping for {self.sleep} seconds...")
            await asyncio.sleep(self.sleep)

    def handle_exit(self, sig, frame):
        self.should_exit = True


def _install_signal_handlers(handle_exit) -> None:
    loop = asyncio.get_event_loop()
    for sig in HANDLED_SIGNALS:
        loop.add_signal_handler(sig, handle_exit, sig, None)


async def main() -> None:
    def _handle_exit(sig, frame):
        server.handle_exit(sig, frame)
        sleeper.handle_exit(sig, frame)
        sleeper_task.cancel()

    server = uvicorn.Server(uvicorn.Config(app))
    sleeper = Sleeper()
    _install_signal_handlers(_handle_exit)
    async with asyncio.TaskGroup() as tg:
        tg.create_task(server._serve(), name="uvicorn")
        sleeper_task = tg.create_task(sleeper.run_forever(), name="sleeper")
    print("DONE")


if __name__ == "__main__":
    asyncio.run(main())

```

**Reply by maxfischer2781** [reply]
This example has the *exact same* bugs in `Sleeper` that #1600 fixed in `uvicorn`. I really don't think recommending to people to roll out their own signal handling is responsible.

**Reply by GianlucaFicarelli** [reply]
@maxfischer2781  could you please clarify why the example above has the same bugs fixed in #1600? It works for me with uvicorn 0.30.1, and it works as well with uvicorn 0.28.1 after:
- replacing `server._serve()` with `server.serve()` in the script above (because `_serve()` didn't exist yet in 0.28.1)
- because I don't want to install the handler, commenting out the line calling `self.install_signal_handlers()` at https://github.com/encode/uvicorn/blob/f73b8beeb1499ca5fcec3067cf89dad5326a0984/uvicorn/server.py#L73


>  I really don't think recommending to people to roll out their own signal handling is responsible

In a complex application (like the ones interested by the fix #1600 as far as I understand), I would probably already need a signal handler, and calling a handler to stop uvicorn would be more explicit than letting it to shutdown by itself IMHO.

Another potential use case: in a GUI, I may want to stop uvicorn by clicking a button, so in that case I need a way to stop uvicorn programmatically (probably without letting uvicorn handling a SIGINT)

EDIT: To be clearer, for the majority of applications where uvicorn is the only task, the current behaviour of handling the signal is appropriate. My suggestion of letting the user handle the signals is meant only for more complex applications, and it shouldn't be the default.

**Reply by maxfischer2781** [reply]
> could you please clarify why the example above has the same bugs fixed in https://github.com/encode/uvicorn/pull/1600?

- The program does not exit with the appropriate killed-by-signal code.
- There is no way to run anything alongside `main` without that being unkillable by CTRL+C.

CTRL+C has well-defined meaning that goes beyond "stop a hardcoded application in some way".

> In a complex application (like the ones interested by the fix https://github.com/encode/uvicorn/pull/1600 as far as I understand), I would probably already need a signal handler, and calling a handler to stop uvicorn would be more explicit than letting it to shutdown by itself IMHO.

No, replacing signal handlers in an `async` library is extremely uncommon. There already is an established way to react to CTRL+C and usually no need to roll out a custom handler.

The issue that triggered me to submit #1600 came absolutely out of the left field for us since custom signal handlers were so unexpected for us in this situation.

> To be clearer, for the majority of applications where uvicorn is the only task, the current behaviour of handling the signal is appropriate. My suggestion of letting the user handle the signals is meant only for more complex applications, and it shouldn't be the default.

No, the pre-#1600 behaviour [is incorrect at least on UNIX](https://www.cons.org/cracauer/sigint.html) and likely on Windows as well (which has a similar mechanism for indicating stopped-by-CTRL+C). [Note that this behaviour was explicitly fixed in Python to establish proper behaviour](https://bugs.python.org/issue1054041). This should absolutely be the default of any production grade piece of software.

**Reply by GianlucaFicarelli** [reply]
> The program does not exit with the appropriate killed-by-signal code.

AFAIK uvicorn (and the script above) can be stopped only by sending a SIGINT/SIGKILL, so it's considered the regular way of exiting (for me it's similar to exit vim with `:q`, for example), and it shouldn't exit with a killed-by-signal code.

The explanation at https://www.cons.org/cracauer/sigint.html makes a lot of sense, but it seems more applicable to programs that are interactive and provide a clean way to exit (like emacs), or that terminate after some work is done (other scripts, for example), rather than servers running forever.

I wonder if there is another way to stop uvicorn cleanly, besides sending SIGINT/SIGKILL?
And what do we expect in a bash script like this, where the uvicorn server is launched internally in myapp.py, and it is stopped by sending SIGINT?

```bash
set -ex
echo start
python myapp.py
echo stop
```

If uvicorn and the script always exit with a killed-by-signal code, there is no way to execute the `echo stop` after it (unless ignoring the exit status).

Similarly, if running `python myapp.py` as the command in a docker container configured with `STOPSIGNAL SIGINT`, then the container will exit with an error after receiving the stop signal, although it acted as intended, and the container exits with status `Exited (130)`.

> There is no way to run anything alongside main without that being unkillable by CTRL+C.

The code in the example above is the full application, so there isn't other code to run alongside main. To add other tasks, it would be needed to add an exit handler for them as well.

It would be different if it was a library, instead. In that case, I would prefer not to install the signal handler, and provide start and stop methods. This is what seems more flexible to me in uvicorn as well, when it needs to be integrated in another app (as a library).

Does this distinction between application and library make sense to you?
